src/net/server/connection.rs

```diff
@@ -0,0 +1,988 @@
+use tokio::sync::mpsc::Sender;
+use tokio::sync::{mpsc, watch};
+use tokio::time::Instant;
+use tokio::time::{sleep_until, timeout};
+use tracing::Level;
+use tracing::{debug, enabled, error, trace, warn};
+
+use crate::base::wire::Composer;
+use crate::base::{Message, StreamTarget};
+use crate::net::server::buf::BufSource;
+use crate::net::server::message::MessageProcessor;
+use crate::net::server::message::Request;
+use crate::net::server::metrics::ServerMetrics;
+use crate::net::server::middleware::chain::MiddlewareChain;
+use crate::net::server::service::{
+    CallResult, Service, ServiceError, ServiceFeedback,
+};
+use crate::net::server::util::to_pcap_text;
+use crate::utils::config::DefMinMax;
+
+use super::message::{
+    MessageDetails, NonUdpTransportContext, TransportSpecificContext,
+};
+use super::middleware::builder::MiddlewareBuilder;
+use super::service::ServerCommand;
+use super::stream::Config as ServerConfig;
+
+/// Limit on the amount of time to allow between client requests.
+///
+/// According to [RFC 7766]:
+/// - "A timeout of at least a few seconds is advisable for normal
+///   operations".
+/// - "Servers MAY use zero timeouts when they are experiencing heavy load or
+///    are under attack".
+/// - "Servers MAY allow idle connections to remain open for longer periods as
+///   resources permit".
+///
+/// The value has to be between zero and 30 days with a default of 30 seconds.
+/// The default and minimum values are the same as those of the Unbound 1.19.2
+/// `tcp-idle-timeout` configuration setting. The upper bound is a guess at
+/// something reasonable.
+///
+/// [RFC 7766]: https://datatracker.ietf.org/doc/html/rfc7766#section-6.2.3
+//
+// Note: Unbound 1.19.2 has another setting, edns-tcp-keepalive-timeout,
+// which if set and edns-tcp-keepalive is set to yes, then Unbound has a
+// default timeout value of 2 minutes instead of 30 seconds. Internally both
+// Unbound options configure the same timeout limit, the tcp-idle-timeout
+// setting may exist for backward compatibility. TO DO: Should we increase
+// the default timeout value to 2 minutes instead of 30 seconds?
+const IDLE_TIMEOUT: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(30),
+    Duration::from_millis(200),
+    Duration::from_secs(30 * 24 * 60 * 60),
+);
+
+/// Limit on the amount of time to wait for writing a response to complete.
+///
+/// The value has to be between 1 millisecond and 1 hour with a default of 30
+/// seconds. These values are guesses at something reasonable. The default is
+/// based on the Unbound 1.19.2 default value for its `tcp-idle-timeout`
+/// setting.
+const RESPONSE_WRITE_TIMEOUT: DefMinMax<Duration> = DefMinMax::new(
+    Duration::from_secs(30),
+    Duration::from_millis(1),
+    Duration::from_secs(60 * 60),
+);
+
+/// Limit on the number of DNS responses queued for writing to the client.
+///
+/// The value has to be between zero and 1,024. The default value is 10. These
```

Comment:
On a related note @partim pointed out that the current implementation does not truly support a streaming response, i.e. that the service should be able to generate data as the client consumes it, and it should. So indeed this is not currently possible with this PR. We'd also like to make bigger changes to the `Service` trait but attempts to change this quickly encountered unforesoon issues, so the conclusion was to look at improving the `Service` trait with one aspect being the way that streaming responses are implemented, but not for this PR. We also discussed that support for zone transfers is not in scope for this PR and so improving this can be part of the next work which will look at transfer support.

Similar to my response above, it isn't clear to me from your comment if you think that (a) the limit should be removed entirely, or (b) that the limit has value for non-XFR transfers and is only an issue for "large" transfers for some value of "large". 

Given these points, please state clearly what change you would like to see here, and which things should be changed iimmediately (in this PR) and which things should be changed but can wait (until after this PR).

Reply:
I've added an item for this to the project backlog and will resolve it here as this relates to XFR which upcoming work will focus on.